'Remove Section' GUI implementation

louob20 opened the pull request:
Appears to work, but please do check that this gives the desired results

I'm also unsure how to format the buttons in a nicer manner

<img width="700" alt="Screenshot 2025-03-10 at 17 53 29" src="https://github.com/user-attachments/assets/17d60ef3-f03b-4d35-a962-fc6e80abee86" />
<img width="697" alt="Screenshot 2025-03-10 at 17 53 48" src="https://github.com/user-attachments/assets/378919bb-dccf-4fbd-ba55-6c8bd0d4c02f" />
<img width="693" alt="Screenshot 2025-03-10 at 17 55 04" src="https://github.com/user-attachments/assets/c48cb906-91e0-4be3-9b65-581da22a33f1" />

aris-mav commented on ext/gui_ext.jl:
```diff
@@ -132,6 +132,7 @@ Run the GUI to plot the 1D inversion results and select peaks you want to label.
 """
 function Makie.plot(res::NMRInversions.inv_out_1D)
 
+    original_res = deepcopy(res)
```
Perhaps we could instead keep the original struct intact, and save a copy of the modified struct. Then we can add a button which replaces the original with the modified one. What do you think?

aris-mav replied:
If so, then the `delete_range!` function can be replaced by a  `delete_range` version, which returns a new structure instead of modifying the existing one in place.

aris-mav commented:
Just tried the code on my machine, it works fine, well done! 
Happy to merge once you consider the points above. 

Also, I was thinking one more thing;
It might be good if we scale the updated distribution after the deletion, so that its integral is the same as the one in the original distribution.
Haven't tested it, but I imagine the fit will be much better after a section is deleted if we do that. Let me know whether you'd be keen to test that.

louob20 commented on src/misc.jl:
```diff
@@ -247,3 +247,24 @@ function weighted_averages(r::inv_out_2D)
+    results.yfit = g
+
+    # pick indices from xfit that are closest to the measured x values 
+    closest_indices = [findmin(abs.(results.xfit .- x))[2] for x in results.x]
```
I've added the original `ker_struct` from the 1D invert method into the results. Calculating the residuals (as in the `finding_alpha` method, using g-Kf with g and K from the kernel struct) with the unchanged `f` leaves the residuals unchanged. However, changing `f` does give seemingly incorrect results. It's likely I'm just misunderstanding how ker_struct is used.


<img width="677" alt="Screenshot 2025-03-11 at 20 20 20" src="https://github.com/user-attachments/assets/ed059174-98d7-4d40-85fe-3e8f02452392" />

<img width="681" alt="Screenshot 2025-03-11 at 20 20 57" src="https://github.com/user-attachments/assets/3855ec8a-9255-4429-b807-8b7d22f7dfbe" />

(even more seemingly incorrect than the last)


The following 2 are the plots of Kf, using the old and new K.

<img width="327" alt="Screenshot 2025-03-11 at 19 44 48" src="https://github.com/user-attachments/assets/2d6ee1c0-3efd-4084-81bf-67d45fc4cac0" />

<img width="332" alt="Screenshot 2025-03-11 at 19 45 47" src="https://github.com/user-attachments/assets/e66babc1-81be-49bf-a671-4e4cf0e3e2b8" />

louob20 commented:
That sounds like a good idea, I will work on this. 

In the current code there is no feature to find the integral of `f` is there?

aris-mav commented:
> That sounds like a good idea, I will work on this.
> 
> In the current code there is no feature to find the integral of `f` is there?

Great.
You're right, trapezoidal intergals would require external packages.
However, we don't necessarily care about the actual integral here.
Something like the sum all the values of the vector would be just as good as the integral in this case. As long as this measure is preserved, it would serve the same purpose as preserving the integral (since the x-axis would be the same before and after).

Perhaps we could also use the `filter` field to encode this information. 
E.g., `filter` is all 1's at the beginning, but once some values are turned into 0's, the remaining 1's can get scaled up accordingly so that `sum(f) == sum( filter .* f)`.
Hope that makes sense.

Of course, let me know if this, or anything else is unclear.

louob20 commented:
Got it, sounds good. Thanks!

louob20 commented:
Hi @aris-mav, I think I've figured out the issue. I have used `create_kernel(res.seq, res.x, res.X) * res.f - res.y` to find the residuals in the end. 

Plotting `create_kernel(res.seq, res.x, res.X) * res.f` gives points that line up perfectly on the fitted curve and align with the res.y values. When removing an empty section, the residuals plot does change, however these must be the correct values.

Looking closely at the first index, having left `f` unchanged:

Original residuals plot,
<img width="402" alt="Screenshot 2025-03-18 at 12 42 36" src="https://github.com/user-attachments/assets/f03de518-5a13-4caa-87e8-9c1a4cf8ce5f" />
Notice how the distance from the point to the fitted curve is ~0.005, but looking at the residual plot it predicts ~0.0001.

Updated residuals plot,
<img width="393" alt="Screenshot 2025-03-18 at 12 43 46" src="https://github.com/user-attachments/assets/52b45ff1-7fba-4474-a628-0123a2b15b17" />
Now the distance from the point to the fitted curve has stayed the same at ~0.005, but the residual plot appears to be much more accurate, reading ~0.0058.

Using the updated method also gives sensible results when removing sections, unlike the original method, where residuals are very large initially and then quickly go to zero.
Additionally, if we plot `K*f` using the previous and updated method, the updated method gives a scatter plot of points arranged exponentially on the fitted curve, whereas the previous method gives odd results:

Original, `res.ker_struct.K * res.f`
<img width="345" alt="Screenshot 2025-03-18 at 13 13 42" src="https://github.com/user-attachments/assets/117db8d1-c2ae-4a74-9f3b-fda83d8d0cc6" />

Updated, `create_kernel(res.seq, res.x, res.X) * res.f`
<img width="330" alt="Screenshot 2025-03-18 at 13 14 40" src="https://github.com/user-attachments/assets/a243c3bb-ca6c-42cb-bd14-f055cd5dc62c" />

louob20 commented:
> That looks good, @louob20! I'll also update how the original residuals are calculated, for consistency.
> 
> Just one thing, as mentioned in a previous comment, it would be good if the `delete_selection!` function only updates the `filter` field of the result structure. Also, `filter_range` might be a more appropriate name for it in this case. This way, we can skip copying the original structure. All the information will be there, and the new residuals/ yfit can be computed just before each plot (for both interactive and non-interactive plots).

So you have to apply the filter to the function only when you want to plot it?

louob20 commented:
@aris-mav I believe that last commit was all, if you would like to run the code before merging

aris-mav commented:
It seems that the filter selection function still updates the residuals and the yfit directly on the structure. 
Am I missing a commit perhaps?

louob20 commented:
I have changed it, should be good now.

aris-mav commented:
Good job @louob20 !
